### DISCUSSION TITLE
Timing issues for pipelined processor and how to debug

### DISCUSSION THREAD

**Discussion by Rayleigh47** [NUS-CG3207/labs#29] [Wrangling Vivado] [answered]
Hi we're trying to run our pipelined processor at 100 Mhz. We noticed that there is a failing path between rs1E and our MCycle's remainder signal. We are not sure why as it seems we are only using rs1E as inputs to the hazard unit and there doesn't seem to be a direct connection between it to our mcycle unit.
![image](https://github.com/user-attachments/assets/32033278-cef8-49b3-9e8d-dc1f0dbfa17e)

Looking a bit deeper, We noticed that our timing requirement is 10ns (since we tried 100mhz clock) but our data delay path is 13+ ns.
![image](https://github.com/user-attachments/assets/67188500-200f-4bfe-b91b-e6d3c367bd2f)

We tried to trace through this data path and we can kinda conclude that our remainder logic is a bit problematic but not really sure where to start debugging at this level.
![image](https://github.com/user-attachments/assets/96638870-4a9c-470c-8398-72e8b2095844)

The signed/unsigned division circuit we use is the basic restoring sequential circuit from the lecture and should take width+1 cycles, is there an issue with our implementation that can cause this?

**Comment by r-panicker** [comment, marked as answer]
I am glad you explored timing. I guess (as you have 32 CARRY4 = 128-bit addition/subtraction = 2 x 64-bit operations that are combinationally related) the problem is that you are doing a test subtraction and restoring in the same cycle. The idea of restoring division is that you test subtract in one cycle and restore it in the next cycle, so that you need only one 64-bit operation per cycle. This will make the number of cycles variable and greater than or equal to 2*width+1 (width=32), but that is what we get with restoring division. Alternatively, have a look at the non-restoring division.

**Reply by Rayleigh47** [reply]
Ok got you Prof, thanks! Will check that out.
